File: external/java/src/main/java/ch/epfl/dedis/lib/omniledger/Argument.java
package ch.epfl.dedis.lib.omniledger;

import ch.epfl.dedis.proto.OmniLedgerProto;
import com.google.protobuf.ByteString;

import java.util.ArrayList;
import java.util.List;

/**
 * Argument is used in an Instruction. It will end up as the input argument of the OmniLedger smart contract.
 */
public class Argument {
    private String name;
    private byte[] value;

    /**
     * Constructor for the argument.
     * @param name The name of the argument.
     * @param value The value of the argument.
     */
    public Argument(String name, byte[] value) {
        this.name = name;
        this.value = value;
    }

    /**
     * Getter for the name.
     * @return The name.
     */
    public String getName() {
        return name;
    }

    /**
     * Getter for the value.
     * @return The value.
     */
    public byte[] getValue() {
        return value;
    }

    /**
     * Converts this object to the protobuf representation.
     * @return The protobuf representation.
     */
    public OmniLedgerProto.Argument toProto() {
        OmniLedgerProto.Argument.Builder b = OmniLedgerProto.Argument.newBuilder();
        b.setName(this.name);
        b.setValue(ByteString.copyFrom(this.value));
        return b.build();
    }

    public static List<Argument> NewList(String key, byte[] value){
        List<Argument> ret = new ArrayList<>();
        ret.add(new Argument(key, value));
        return ret;
    }

    public static List<Argument> NewList(String key, String value) {
        return NewList(key, value.getBytes());
    }
}

File: external/js/cothority/lib/omniledger/Argument.js
class Argument {
  /**
   * Argument is used in an Instruction. It will end up as the input argument
   * of the OmniLedger smart contract.
   * @param {string} name - The name of the argument
   * @param {Uint8Array} value - The value of the argument
   */
  constructor(name, value) {
    this._name = name;
    this._value = value.slice(0);
  }

  /**
   * Getter for the name
   * @return {string}
   */
  get name() {
    return this._name;
  }

  /**
   * Getter for the value
   * @return {Uint8Array}
   */
  get value() {
    return this._value.slice(0);
  }

  /**
   * Create an object with all the necessary field needed to be a valid message
   * in the sense of protobufjs. This object can then be used with the "create"
   * method of protobuf
   *
   * @return {Object}
   */
  toProtobufValidMessage() {
    return {
      name: this.name,
      value: this.value
    };
  }
}

module.exports = Argument;
